python/lsst/ip/diffim/imageDecorrelation.py

```diff
@@ -196,235 +196,239 @@ def run(self, exposure, templateExposure, subtractedExposure, psfMatchingKernel,
-        out = kernel
-        changed = False
-        if (out.shape[0] % 2) == 1:
-            out = np.pad(out, ((1, 0), (0, 0)), mode='constant')
-            changed = True
-        if (out.shape[1] % 2) == 1:
-            out = np.pad(out, ((0, 0), (1, 0)), mode='constant')
-            changed = True
-        if changed:
-            out *= (np.mean(kernel) / np.mean(out))  # need to re-scale to same mean for FFT
-        return out
+        kappa = self.padCenterOriginArray(kappa, self.freqSpaceShape)
+        kft = np.fft.fft2(kappa)
+        kft2 = np.conj(kft) * kft
+        if preConvArr is None:
+            denom = svar + tvar * kft2
+        else:
+            preConvArr = self.padCenterOriginArray(preConvArr, self.freqSpaceShape)
+            mk = np.fft.fft2(preConvArr)
+            mk2 = np.conj(mk) * mk
+            denom = svar * mk2 + tvar * kft2
+        kft = np.sqrt((svar + tvar) / denom)
+        return kft
 
-    @staticmethod
-    def _fixEvenKernel(kernel):
-        """Take a kernel with even dimensions and make them odd, centered correctly.
+    def computeCorrectedDiffimPsf(self, corrft, psfArr):
+        """Compute the (decorrelated) difference image's new PSF.
 
         Parameters
         ----------
-        kernel : `numpy.array`
-            a numpy.array
+        corrft : `numpy.ndarray`
+            The frequency space representation of the correction calculated by
+            `computeCorrection`. Shape must be `self.freqSpaceShape`.
+        psfArr : `numpy.ndarray`
+            The psf of the difference image to be corrected.
 
         Returns
         -------
-        out : `numpy.array`
-            a fixed kernel numpy.array
-        """
-        # Make sure the peak (close to a delta-function) is in the center!
-        maxloc = np.unravel_index(np.argmax(kernel), kernel.shape)
-        out = np.roll(kernel, kernel.shape[0]//2 - maxloc[0], axis=0)
-        out = np.roll(out, out.shape[1]//2 - maxloc[1], axis=1)
-        # Make sure it is odd-dimensioned by trimming it.
-        if (out.shape[0] % 2) == 0:
-            maxloc = np.unravel_index(np.argmax(out), out.shape)
-            if out.shape[0] - maxloc[0] > maxloc[0]:
-                out = out[:-1, :]
-            else:
-                out = out[1:, :]
-            if out.shape[1] - maxloc[1] > maxloc[1]:
-                out = out[:, :-1]
-            else:
-                out = out[:, 1:]
-        return out
+        psfNew : `numpy.ndarray`
+            The corrected psf, same shape as `psfArr`, sum normed to 1.
 
-    @staticmethod
-    def _doConvolve(exposure, kernel):
-        """Convolve an Exposure with a decorrelation convolution kernel.
+        Notes
+        ----
+        There is no algorithmic guarantee that the corrected psf can
+        meaningfully fit to the same size as the original one.
+        """
+        psfShape = psfArr.shape
+        psfArr = self.padCenterOriginArray(psfArr, self.freqSpaceShape)
```

Comment:
I think you want to use `psfNew = self.padCenterOriginArray(psfArr, self.freqSpaceShape)` here, and then use `psfNew` for the rest of the calculations.